src/ObjCRuntime/Runtime.cs

```diff
@@ -1684,12 +1686,26 @@ static Type LookupINativeObjectImplementation (IntPtr ptr, Type target_type, Typ
 					// native objects and NSObject instances.
 					throw ErrorHelper.CreateError (8004, $"Cannot create an instance of {implementation.FullName} for the native object 0x{ptr:x} (of type '{Class.class_getName (Class.GetClassForObject (ptr))}'), because another instance already exists for this native object (of type {o.GetType ().FullName}).");
 				}
-				return (T?) ConstructNSObject<T> (ptr, implementation, MissingCtorResolution.ThrowConstructor1NotFound);
+				var rv = (T?) ConstructNSObject<T> (ptr, implementation, MissingCtorResolution.ThrowConstructor1NotFound);
+				if (owns)
+					TryReleaseINativeObject (rv);
+				return rv;
 			}
 
 			return ConstructINativeObject<T> (ptr, owns, implementation, MissingCtorResolution.ThrowConstructor2NotFound);
 		}
 
+		static void TryReleaseINativeObject (INativeObject? obj)
+		{
+			if (obj is null)
+				return;
+
+			if (!(obj is NativeObject nobj))
+				throw new InvalidOperationException ($"Unable to call release on an instance of the type {obj.GetType ()}");
```

Comment:
Should this be in its own method, so the JIT can inline TryReleaseINativeObject?

Reply:
I don't think this is a hot codepath, because *a lot* of other things are also executed around when this method is called.